Orange/classification/rules.py

```diff
@@ -0,0 +1,938 @@
+import operator
+from copy import copy
+from hashlib import sha1
+from abc import ABCMeta, abstractmethod
+
+import numpy as np
+from scipy.stats import chisqprob
+import Orange
+from Orange.classification import Learner, Model
+
+__all__ = ["CN2Learner", "CN2UnorderedLearner"]
+
+
+def argmaxrnd(a, random_seed=None):
+    """
+    Find the index of the maximum value for a given 1-D numpy array.
+
+    In case of multiple indices corresponding to the maximum value, the
+    result is chosen randomly among those. The random number generator
+    can be seeded by forwarding a seed value: see function 'hash_dist'.
+
+    Parameters
+    ----------
+    a : array_like
+        The source array.
+    random_seed : int
+        RNG seed.
+
+    Returns
+    -------
+    index : int
+        Index of the maximum value.
+
+    Raises
+    ------
+    ValueError : shape mismatch
+        If 'a' has got more than 2 dimensions.
+
+    Notes
+    -----
+    2-D arrays are also supported to avoid multiple RNG initialisation.
+    An array of indices corresponding to the maximum value of each row
+    is then returned.
+    """
+    if a.ndim > 2:
+        raise ValueError("argmaxrnd only accepts arrays of up to 2 dim")
+
+    random = (np.random if random_seed is None
+              else np.random.RandomState(random_seed))
+
+    def f(x): return random.choice((x == np.nanmax(x)).nonzero()[0])
+    return f(a) if a.ndim == 1 else np.apply_along_axis(f, axis=1, arr=a)
+
+
+def entropy(x):
+    x = x[x != 0]
+    x = x / np.sum(x, axis=0)
+    x = -x * np.log2(x)
+    return np.sum(x)
+
+
+def likelihood_ratio_statistic(x, y):
+    x[x == 0] = 1e-5
+    y[y == 0] = 1e-5
+    lrs = np.sum(x * np.log(x / y))
+    lrs = 2 * (lrs - np.sum(x) * np.log(np.sum(x) / np.sum(y)))
```

Comment:
If `x` and `y` are arrays, which they are`, you can call `.sum()` methods directly, which is marginally faster and shorter to write.